quinn-proto/src/config.rs

```diff
@@ -838,6 +846,54 @@ impl ServerConfig {
         self.preferred_address_v6 = address;
         self
     }
+
+    /// Maximum number of [`Incoming`][crate::Incoming] to allow to exist at a time
+    ///
+    /// An [`Incoming`][crate::Incoming] comes into existence when an incoming connection attempt
+    /// is received and stops existing when the application either accepts it or otherwise disposes
+    /// of it. While this limit is reached, new incoming connection attempts are immediately
+    /// refused. Larger values have greater worst-case memory consumption, but accommodate greater
+    /// application latency in handling incoming connection attempts.
+    ///
+    /// The default value is set to 65536. Assuming each [`Incoming`][crate::Incoming] accounts for
+    /// little over 1200 bytes of memory at most, this should limit memory consumption from this to
+    /// under 100 MiB--a generous amount that still prevents memory exhaustion in most contexts.
+    pub fn max_incoming(&mut self, max_incoming: usize) -> &mut Self {
+        self.max_incoming = max_incoming;
+        self
+    }
+
+    /// Maximum number of received bytes to buffer for each [`Incoming`][crate::Incoming]
+    ///
+    /// An [`Incoming`][crate::Incoming] comes into existence when an incoming connection attempt
+    /// is received and stops existing when the application either accepts it or otherwise disposes
+    /// of it. This limit governs only packets received within that period, and does not include
+    /// the first packet. Packets received in excess of this limit are dropped, which may cause
+    /// 0-RTT data to have to be retransmitted.
+    ///
+    /// The default value is set to 10 MiB--an amount such that in most situations a client would
+    /// not transmit that much 0-RTT data faster than the server handles the corresponding
+    /// [`Incoming`][crate::Incoming].
+    pub fn incoming_buffer_size(&mut self, incoming_buffer_size: u64) -> &mut Self {
```

Comment:
Do we need/want this? Seems to me most servers will mostly/only care about `_total`?

Reply:
I'm not sure. I might be off on this comparison, but it seems abstractly analogous to the transport parameters having an initial max data limit both per each stream and for all streams in total.

Earlier, before concerns around interaction with raw datagrams were raised, I was computing the default for this limit based on the maximum valid amount of stream data bytes a 0-rtt client could transmit.

The scenarios I'm trying to imagine where it's useful _are_ seeming rather esoteric. I wonder what @Ralith thinks about whether this limit is useful enough to justify its existence?

Reply:
Having separate limits is desirable in principle because it enables better behavior when a single badly behaved connection is saturating the buffer. With a single global limit, it's easier for an attacker to cause 0-RTT or initial packets for other connections to be lost, even when the total limit is the same.

Maybe offering a separate knob is overkill in this case. Well-behaved servers should consume `Incoming`s in a timely manner, so the period for which any single badly behaved connection can saturate the buffer is short. We could leave this off for simplicity's sake in the absence of specific demand; I don't have a strong opinion at the moment.

Reply:
Okay, given that we (mostly @gretchenfrage!) have already done the work and that additional complexity is probably limited, let's keep it for now.